Load Testing Concept (#3796)
kcp-dev/kcp: test/load/Readme.md

# Load Testing

## 10000 workspaces reference architecture

Assumptions & Requirements:

* We want to test how a kcp installation with 10000 workspaces behaves on simulated, regular

> activities. We don't want to test how easily we can add 10000 workspaces at once

Reviewer: ```suggestion
* We want to test how a kcp installation with 10000 workspaces behaves on synthetic workloads. We don't want to test how easily we can add 10000 workspaces at once
```

> * You don't want to run etcd on machines which are hosting kcp

Reviewer: ```suggestion
* We don't want to run etcd on machines which are hosting kcp
```

* We will be using kcp operator to setup a sharded kcp instance
* The minimum amount of replicas for any group is 3
* The loadtests are infrastructure agnostic. Yes your kcp needs to have the apporpriate size, but
the test does not require any specific architecture. This will allow us and community members to

> experiment with different architectures

Reviewer: ```suggestion
* The loadtests are platform agnostic and only rely on Kubernetes compute
```

I think the appropriate sizing is clear. Architecture for me means arm or x86 - I suppose this point means that it doesn't require the use of a specific vendor (AWS, BTP) but instead only requires a set of kubeconfigs to connect to kube compute.

  ```

#### Example for 10000 workspaces

```txt
#shards = 10000 / 3500 = 2,85 = 3
#etcd_nodes = 3 * 3 = 9
#shard_nodes = 3 * 3 = 9
#actual_workspaces_per_shard = 10000 / 3 = 3333
kcp_server_node_mem = 512 + (3333 * 5) = 17777MB
#total_nodes =  3 + 1 + 9 + 3 + 1 = 17
```

### Testing Protocol

Mantra: We want to test how a kcp installation with 10000 workspaces behaves on simulated, regular
activities. We don't want to test how easily we can add 10000 workspaces at once

#### Procedure

1. Create 10000 workspaces, APIExports, etc. and patiently wait for all of them to become ready
2. Simulate real workd activity by simulating end-users using custom kubeconfigs to create APIBindings
and then CRUD on their custom api-objects

##### Level 1 - 10000 empty workspaces

We are just going to put 10000 empty workspaces into a kcp installation. This is extremely deterministic and has litte randomness. We mainly use this for a base consumption measurement.

##### Level 2 - Basic CRUD

Every workspace has a type and we are going to do a basic parallel CRUD workflow which we will simulate
using simple Cronjobs. For the CRUD flow we anticipate

> The workflow is done on basic objects from a single provider using a singular APIExport.

Reviewer 2: the sentence can continue after:
"For the CRUD flow we anticipate"

##### Level 3 - Multiple Providers

We are multixplexing the level 2 example to use multiple providers.

##### Outlook

We want to keep the initial version of the tests simple and deterministic. As a result the following

> following topics have been discussed, but were considered not to be part of the first 3 level implementation:

Reviewer 2: ```suggestion
We want to keep the initial version of the tests simple and deterministic. As a result the following
topics have been discussed, but were considered not to be part of the first 3 level implementation:
```
